### PR TITLE
add the `voxpopuli` recipe

### PR DESCRIPTION
**Pull request by vesis84**
- this is the data preparation
- there is no ASR training and no results

### CODE REVIEW

**Comment by vesis84** [comment]
Hi, who would review this ?
K.

File: egs/voxpopuli/ASR/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Readme
+
+This recipe contains data preparation for the VoxPopuli dataset.
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/README.md:3]: Could you provide a hyperlink for VoxPopuli?
E.g., use something like below
```
[VoxPopuli](http://some/ulr/for/VoxPopuli)
```

File: egs/voxpopuli/ASR/local/compute_fbank.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+      --speed-perturb True
+
+It looks for raw CutSet in the directory data/fbank
+located at: `{src_dir}/{prefix}_cuts_{dataset}_raw.jsonl.gz`.
+
+The generated fbank features are saved in `data/fbank/{prefix}-{dataset}_feats`
+and CutSet manifest stored in `data/fbank/{prefix}_cuts_{dataset}.jsonl.gz`.
+
+The number of workers is smaller than nunber of jobs
+(see --num-jobs 100 --num-workers 25 in the example).
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/compute_fbank.py:40]: Is it a hard constraint? If so, could you write a check in the code for it?

File: egs/voxpopuli/ASR/local/preprocess_voxpopuli.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
+# Copyright    2023  Xiaomi Corp.        (authors: Yifan Yang)
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/preprocess_voxpopuli.py:2]: Maybe you need to update the author information?

File: egs/voxpopuli/ASR/local/separate_punctuation.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
+#!/bin/env python3
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/separate_punctuation.py:1]:
```suggestion
#!/usr/bin/env python3
```

File: egs/voxpopuli/ASR/local/separate_punctuation.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ but that would be an extra dependency)
+
+It can be used as a module, or as an executable script.
+
+Usage example #1:
+  from separate_punctuation import separate_punctuation
+
+Usage example #2:
+  python3 ./local/separate_punctuation.py \
+    --ignore-columnts 1 \
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/separate_punctuation.py:35]:
```suggestion
    --ignore-columns 1 \
```

File: egs/voxpopuli/ASR/local/separate_punctuation.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ${kaldi_data}/text
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/separate_punctuation.py:36]: This code seems to read the content from `stdin`, not the provided file `${kaldi_data}/text`. Is it correct?

File: egs/voxpopuli/ASR/local/text_from_manifest.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+#!/bin/env python3
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/text_from_manifest.py:1]:
```suggestion
#!/usr/bin/env python3
```

File: egs/voxpopuli/ASR/local/uppercase_begin_of_sentence.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
+#!/bin/env python3
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/uppercase_begin_of_sentence.py:1]:
```suggestion
#!/usr/bin/env python3
```

File: egs/voxpopuli/ASR/local/uppercase_begin_of_sentence.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+"""
+This script introduces initial capital letter at the beginning of a sentence.
+It can be used as a module, or as an executable script.
+
+Usage example #1:
+  from uppercase_begin_of_sentence import UpperCaseBeginOfSentence
+
+Usage example #2:
+  python3 ./local/uppercase_begin_of_sentence.py \
+    --ignore-columnts 1 \
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/uppercase_begin_of_sentence.py:26]:
```suggestion
    --ignore-columns 1 \
```

File: egs/voxpopuli/ASR/local/validate_cutset_manifest.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/local/validate_cutset_manifest.py:1]: Could you replace this file with a symlink to its original one if no changes are done to this file?

File: egs/voxpopuli/ASR/prepare.sh [hunk 11]
```diff
@@ -0,0 +1,263 @@
+#!/usr/bin/env bash
+
+. /mnt/matylda5/iveselyk/ASR_TOOLKITS/K2_SHERPA_PYTORCH20/conda-activate.sh
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/prepare.sh:3]: Please add the following two lines
https://github.com/k2-fsa/icefall/blob/6d275ddf9fdd67a32b79b93d70fedffe4b156d5c/egs/librispeech/ASR/prepare.sh#L3-L4

File: egs/voxpopuli/ASR/prepare.sh [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+#musan_dir=${dl_dir}/musan
+musan_dir=/mnt/matylda2/data/MUSAN
+
+# Choose vlues from:
+#
+# "en", "de", "fr", "es", "pl", "it", "ro", "hu", "cs", "nl", "fi", "hr",
+# "sk", "sl", "et", "lt", "pt", "bg", "el", "lv", "mt", "sv", "da",
+# "asr", "10k", "100k", "400k"
+#
+# See: https://github.com/lhotse-speech/lhotse/blob/c5f26afd100885b86e4244eeb33ca1986f3fa923/lhotse/bin/modes/recipes/voxpopuli.py#L77
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/prepare.sh:54]: This URL links to
```
@click.option(
    "--subset",
    type=click.Choice(["asr", "10k", "100k", "400k"] + LANGUAGES + LANGUAGES_V2),
    default="asr",
)
```

Should it be associated with the next line
```
task=asr
```
?

File: egs/voxpopuli/ASR/prepare.sh [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+          --prefix "voxpopuli-${task}-${lang}" \
+          --dataset ${dataset} \
+          --trim-to-supervisions True
+      touch data/fbank/.voxpopuli-${task}-${lang}-${dataset}.done
+    fi
+  done
+fi
+
+if [ $stage -le 5 ] && [ $stop_stage -ge 5 ]; then
+  log "Stage 6: Compute fbank for train set of VoxPopuli"
```
Comment (csukuangfj) [inline egs/voxpopuli/ASR/prepare.sh:163]:
```suggestion
  log "Stage 5: Compute fbank for train set of VoxPopuli"
```

**Comment by csukuangfj** [comment]
> Hi, who would review this ? K.

Sorry for the late reply.

Just left some minor comments. Otherwise, everything looks good to me.

File: egs/voxpopuli/ASR/local/compute_fbank.py (hunk 2 above)
Comment (vesis84) [reply egs/voxpopuli/ASR/local/compute_fbank.py:40]: it is not a hard constraint, there can be also less jobs  than workers, in this case some workers will be idle...

File: egs/voxpopuli/ASR/local/validate_cutset_manifest.py (hunk 10 above)
Comment (vesis84) [reply egs/voxpopuli/ASR/local/validate_cutset_manifest.py:1]: Actually, it is extended version of `validate_manifest.py` from Librispeech. Should I change it directly in the Librispeech recipe ? (for now i'll update it in the VoxPopuli recipe, and i'll add validation to `prepare.sh` too)

**Comment by vesis84** [comment]
Hi @csukuangfj , thank you for the review. The issues should be fixed now.
I'll re-run the data preparation to see if it's okay.
K.

**Comment by csukuangfj** [comment]
> Hi @csukuangfj , thank you for the review. The issues should be fixed now. I'll re-run the data preparation to see if it's okay. K.

Thanks!

Looks good to me. Please leave a message when you think it is ready to merge.

File: egs/voxpopuli/ASR/local/preprocess_voxpopuli.py (hunk 3 above)
Comment (vesis84) [reply egs/voxpopuli/ASR/local/preprocess_voxpopuli.py:2]: ok, done here and also in other .py scripts...

File: egs/voxpopuli/ASR/local/compute_fbank.py (hunk 2 above)
Comment (vesis84) [reply egs/voxpopuli/ASR/local/compute_fbank.py:40]: ok, the assert is added into the code

**Comment by vesis84** [comment]
Hi @csukuangfj ,
to me it seems done. I fixed or explained all the issues, fixed quite some typos too. 
I did a re-run of the data preparation here locally.

Please, have a look one more time, if possible.
Best regards,
Karel

**Comment by csukuangfj** [comment]
@vesis84 

Thanks for your contribution!

**Pull request merged**